### PR TITLE
Wip enr

### PR DESCRIPTION
**Pull request by vanessabridge**
ENR with nodes entering and exiting the network
changes to ProgresPerTime to update list of LiveNodes

### CODE REVIEW

File: src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java [hunk 1]
```diff
@@ -76,6 +76,7 @@ public void run(Predicate<Protocol> contIf) {
         if (p.network().time % 10000 == 0) {
           System.out.println("time goes by... time=" + (p.network().time / 1000) + ", stats=" + s);
         }
+        liveNodes = p.network().allNodes.stream().filter(n -> !n.down).collect(Collectors.toList());
       } while (contIf.test(p));
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java]: WOuldn't be better to recalculate the liveNodes just after the 'p.network().runMs(10);' above?

File: src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java [hunk 2]
```diff
@@ -89,6 +90,10 @@ public void run(Predicate<Protocol> contIf) {
       System.out.println("msg rcvd: " + StatsHelper.getStatsOn(liveNodes, Node::getMsgReceived));
       System.out.println("done at: " + StatsHelper.getStatsOn(liveNodes, Node::getDoneAt));
       System.out.println("Simulation execution time: " + ((endAt - startAt) / 1000) + "s");
+      System.out.println("");
+      System.out.println("Number of nodes that are down"
+          + p.network().allNodes.stream().filter(n -> ((Node) n).down == true).count());
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java]: The cast (Node) is not needed, no?

File: src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java [hunk 3]
```diff
@@ -130,13 +130,31 @@ private void selectChangingNodes() {
...
+  }
+
   @Override
   public void init() {
     for (int i = 0; i < params.NODES; i++) {
       network.addNode(new ETHNode(network.rd, this.nb, generateCap()));
     }
     network.setPeers();
 
+    selectChangingNodes();
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java]: In selectChangingNodes we can select nodes that are down. This seems wrong.

File: src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java [hunk 4]
```diff
@@ -338,10 +360,11 @@ private void capSearch() {
 
       @Override
       public StatsHelper.Stat get(List<? extends Node> liveNodes) {
-        return new StatsHelper.Counter(liveNodes.stream().filter(n -> n.getDoneAt() > 0).count());
+        return new StatsHelper.Counter(
+            liveNodes.stream().filter(n -> n.getDoneAt() > 0 && !n.down).count());
       }
     };
-
+    System.out.println("nodes that have left: ");
```
Comment (nkeywal) [inline src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java]: I guess we should remove all the println stuff...

File: src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java (hunk 3 above)
Comment (vanessabridge) [reply src/main/java/net/consensys/wittgenstein/protocols/ENRGossiping.java]: as we select nodes that change parameters prior to running (during the initial setup stage) all nodes are still up at that time.

File: src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java (hunk 1 above)
Comment (vanessabridge) [reply src/main/java/net/consensys/wittgenstein/core/ProgressPerTime.java]: moved

**Comment by vanessabridge** [comment]
ENR should be ready to run, there is some documentation needed to talk about certain parameters and testing is also needed

**Pull request merged**